lib/pure/options.nim

```diff
@@ -176,21 +310,239 @@ proc filter*[T](self: Option[T], callback: proc (input: T): bool): Option[T] =
+  var caseStmt = nnkCaseStmt.newTree(
+    value
+  )
+  for c in cases:
+    var ofBranch = nnkOfBranch.newTree(
+      c.check,
+      newStmtList()
+    )
+    if c.ident != nil:
+      ofBranch[1].add newLetStmt(c.ident, value)
+    ofBranch[1].add c.body
+    caseStmt.add ofBranch
+  if elseCase.body != nil:
+    if elseCase.ident != nil:
+      caseStmt.add nnkElse.newTree(newStmtList(
+        newLetStmt(elseCase.ident, value),
+        elseCase.body
+      ))
+    else:
+      caseStmt.add nnkElse.newTree(elseCase.body)
+  result = quote do:
+    (proc (): auto =
+      let `optionSym` = `option`
+      if `optionSym`.isSome:
+        `caseStmt`
+      else:
+        `noneCase`
+    )()
+  #echo result.repr
+
+proc optCmp*[T](self: Option[T], value: Option[T],
+  cmp: proc (val1, val2: T): bool): Option[T] =
+  ## Comparison operation that return a some iff ``self`` is a some and the
+  ## value of ``cmp`` applied to ``self`` and ``value`` returns true. The
+  ## returned some has the value of ``self``.
+  if self.isSome and value.isSome and cmp(self.val, value.val):
+    return self
+
+template `==`*[T](self: Option[T], value: Option[T]): Option[T] =
+  ## Wrapper for optCmp with ``==`` as the comparator
+  optCmp(self, value, proc (val1, val2: T): bool = val1 == val2)
+
+template `not`*[T](self: Option[T]): Option[T] =
+  ## Not always returns a ``none`` as a ``none`` can't implicitly get a value
+  none[T]()
+
+template `<=`*[T](self: Option[T], value: Option[T]): Option[T] =
+  ## Wrapper for optCmp with ``<=`` as the comparator
+  optCmp(self, value, proc (val1, val2: T): bool = val1 <= val2)
+
+template `>=`*[T](self: Option[T], value: Option[T]): Option[T] =
+  ## Wrapper for optCmp with ``>=`` as the comparator
+  optCmp(self, value, proc (val1, val2: T): bool = val1 >= val2)
+
+template `>`*[T](self: Option[T], value: Option[T]): Option[T] =
+  ## Wrapper for optCmp with ``>`` as the comparator
+  optCmp(self, value, proc (val1, val2: T): bool = val1 > val2)
+
+template `<`*[T](self: Option[T], value: Option[T]): Option[T] =
+  ## Wrapper for optCmp with ``<`` as the comparator
+  optCmp(self, value, proc (val1, val2: T): bool = val1 < val2)
+
+proc toOpt*[T](value: Option[T]): Option[T] =
+  ## Procedure with overload to automatically convert something to an option if
+  ## it's not already an option.
+  return value
+
+proc toOpt*[T](value: T): Option[T] =
+  ## Procedure with overload to automatically convert something to an option if
+  ## it's not already an option.
+  some(value)
+
+template `or`*[T](x, y: Option[T]): Option[T] =
+  ## When ``x`` is none, return ``y``
+  let xx = x
+  if xx.isSome:
+    xx
+  else:
+    y
+
+template `and`*[T,U](x: Option[T]; y: Option[U]): Option[U] =
+  ## When ``x`` is some, then return ``y``, else return ``none``
+  if x.isSome:
+    y
+  else:
+    none[U]()
+
+converter toSome*[T](value: T): Option[T] =
+  ## Automatic wrapping of values into ``some``
+  some(value)
```

Comment:
I don't think using converters is a good idea here. `T` and `option[T]` are different for good reasons, I prefer excplitly writing `some(x)` if necessary

Reply:
Yeah I wasn't sure about that either and it seems to already cause some confusion.. The idea was to reduce the effort of using options but I guess it might be better to err on the side of being explicit here.